Mathlib/Data/List/Defs.lean

```diff
@@ -749,4 +749,87 @@ Example: if `f : ℕ → list ℕ → β`, `List.mapWithComplement f [1, 2, 3]`
 def mapWithComplement {α β} (f : α → List α → β) : List α → List β :=
   mapWithPrefixSuffix fun pref a suff => f a (pref ++ suff)
 
+/--
+`List.choose` filters list `l` of type `List α` with `chooser` which is of type `α → Option β`,
+yet `β` is ignored.
+It keeps the elements where `chooser x` is `some _` and discards the ones that are `none`
+
+Example:
+```
+def oneTwo? : Nat → Option Nat
+  | 1 => some 1
+  | 2 => some 2
+  | _ => none
+
+[1, 2, 3, 5, 1, 4, 2].choose oneTwo? = [1, 2, 1, 2]
+```
+-/
+def choose (chooser : α → Option β) (l : List α) : List α :=
+  l.filter (Option.isSome ∘ chooser)
+
+/--
+Auxiliary function for `List.compareWith`
+-/
+def compareWithAux (comparer : α → α → Int) : List α → List α → Int
+  | x::xs, y::ys =>
+    let c := comparer x y
+    if c = 0 then
+      compareWithAux comparer xs ys
+    else c
+  | [], [] => 0
+  | _, [] => 1
+  | [], _ => -1
+
+/--
+`List.compareWith` compares two lists `l₁` and `l₂` with the given function `comparer`,
+element by element.
+Example:
+```
+def compare' (x y : Int) : Int :=
+  if x = y then 0
+  else
+    if x > y then 1
+    else -1
+
+[1, 5].compareWith compare' [1, 8] = -1
+[1, 10].compareWith compare' [1, 10] = 0
+[1, 11].compareWith compare' [1, 13] = 1
+[1, 5].compareWith compare' [1] = -1
+[1].compareWith compare' [1, 8] = 1
+```
+-/
+def compareWith (comparer : α → α → Int) (l₁ l₂ : List α) : Int :=
+  compareWithAux comparer l₁ l₂
+
+/--
+`List.sum` sums all of the elements of of list `l`, whose elements
+have instances of `Add` and `OfNat`.
+-/
+def sum [Add α] [OfNat α 0] (l : List α) : α :=
+  l.foldl (.+.) 0
+
+/--
+`List.sumBy` sums the results of `f` mapped to each element of `l`. The return type of `f`
+must haave instances of `Add` and `OfNat`.
+-/
+def sumBy [Add β] [OfNat β 0] (l : List α) (f : α → β) : β :=
+  l.map f |>.sum
+
+/--
+`List.average` computes the average of all the elements in the
+given list.
+-/
+def average [Add α] [HDiv α Nat α] [OfNat α 0] : List α -> α
```

Comment:
The `HDiv α Nat α` constraint will probably only be satisfied by `α = Nat`, but this should work for now.

Reply:
I had a typeclass instance that fixed that, but I forgot to add it in

Reply:
I will add it back

Reply:
Fixed.